examples/typing/TypeSystem.jl

```diff
@@ -0,0 +1,79 @@
+# -*- coding: utf-8 -*-
+# ---
+# jupyter:
+#   jupytext:
+#     text_representation:
+#       extension: .jl
+#       format_name: light
+#       format_version: '1.3'
+#       jupytext_version: 0.8.6
+#   kernelspec:
+#     display_name: Julia 1.0.3
+#     language: julia
+#     name: julia-1.0
+# ---
+
+using Cassette;
+using DifferentialEquations;
+
+# +
+# define the context we will use
+ctx = Cassette.@context typCtx
+
+# add boilerplate for functionality
+function Cassette.overdub(ctx::typCtx, args...)
+    println(typeof.(args))
+    if Cassette.canrecurse(ctx, args...)
+        newctx = Cassette.similarcontext(ctx, metadata = ctx.metadata)
+        return Cassette.recurse(newctx, args...)
```

Comment:
can we get the return type from this line?

Reply:
I can add the logic to return type, but  as a reminder for ode functions, they are returned in place. I will add a try and catch loop

Reply:
oh yeah I forgot about that. We need a way to represent inplace modification. For example the du/dt part are modified but the  parameters and u and t are not. Maybe we need to do more analysis to detect that. We can record 
```julia
oldhash = hash.(args)
out = f(args...)
newhash = hash.(args)
```
and then do `any(.==(oldhash, newhash))` to detect the arguments that change.
Then do a union over all calls to create the set that is effective.

Reply:
I think that instead of that I know that we can access the results with <stuct>.(fields). I think it would make more sense to add an iterator on thoses (ie using getfields or methods)

Reply:
@jpfairbanks I realized something. If we are looking at ODEs specifically we should just do this 
![image](https://user-images.githubusercontent.com/41996797/52713202-ac921280-2f64-11e9-8473-0133e74bbfab.png) 
The other thing to consider is that this is not actually the hashed function since the solve(sir_prob) where the calculations are done is the real output.

Reply:
oh there is a `setfield!/getfield` pair. so you can see if a function modifies a variable by calling `setfield!`.